Add undetermined state to wxCheckBoxXmlHandler

Randalphwa opened the pull request:
A 3-state checkbox can be initialized as `checked` or `undetermined` -- this adds support for the `undetermined` initial state, but only if the checkbox style is set to wxCHK_3STATE.

Randalphwa committed:
Add undetermined state to wxCheckBoxXmlHandler

3-state checkboxes can have their initial value set to checked
or undetermined -- this adds support for the undetermined
state to the handler (ignored if style is not set to wxCHK_3STATE.

Randalphwa committed:
Add undetermined state to wxCheckBoxXmlHandler

3-state checkboxes can have their initial value set to checked
or undetermined -- this adds support for the undetermined
state to the handler (ignored if style is not set to wxCHK_3STATE.

vadz commented on docs/doxygen/overviews/xrc_format.h:
```diff
@@ -937,6 +937,8 @@ No additional properties.
 @row3col{checked, @ref overview_xrcformat_type_bool,
      Should the checkbox be checked initially (default: 0)?}
+@row3col{undetermined, @ref overview_xrcformat_type_bool,
+     Should a 3-state checkbox be initialized as undetermined (default: 0)?}
```
Could you please add `@since 3.1.7`?

vadz commented on src/xrc/xh_chckb.cpp:
```diff
@@ -43,7 +43,11 @@ wxObject *wxCheckBoxXmlHandler::DoCreateResource()
-    control->SetValue(GetBool( wxT("checked")));
+    if (GetBool("checked"))
+        control->SetValue(true);
+    else if (GetBool("undetermined") && (GetStyle() & wxCHK_3STATE))
+        control->Set3StateValue(wxCHK_UNDETERMINED);
```
I think we should report an error if `undetermined` is used for a checkbox without 3 state style, it would be IMO more helpful than silently ignoring it. There is `ReportError()` which could be used for this.

What do you think?

vadz commented on misc/schema/xrc_schema.rnc:
```diff
@@ -760,7 +760,8 @@ wxCheckBox =
         [xrc:p="important"] element label {_, t_text }* &
-        [xrc:p="o"] element checked {_, t_bool }*
+        [xrc:p="o"] element checked {_, t_bool }* &
+        [xrc:p="o"] element undetermined {_, t_bool }*
```
I wonder if we should be really using `bool` for this one. This allows using `<undetermined>false</undetermined>` which is rather strange. Perhaps we should extend `checked` to allow it taking some other value? The trouble is that I don't know which value would fit with the currently allowed `0` and `1` values, neither `-1` nor `2` seem like a good fit.

vadz replied:
We need to take some decision here if we want to have this in 3.1.7, so let's use 0, 1 and 2 for the states if nobody objects too much.

TIA!

vadz reviewed (commented):
Thanks, we definitely want to have this but I'm not totally sure if using a new attribute is the best way to do it. Does anybody else have any thoughts about this by chance?

And, in any case, I think we should report an error if an attempt to set the 3rd state is made for a 2-state checkbox.

Randalphwa committed:
Update to use checked: 0, 1, 2

This turns the checked property from a bool into an integer,
and supports 2 for setting the undetermined state of a
3-state checkbox.

Randalphwa committed:
Merge branch 'wxWidgets:master' into xrc_checkbox

Randalphwa committed:
Merge branch 'xrc_checkbox' of https://github.com/Randalphwa/wxWidgets into xrc_checkbox

Randalphwa committed:
Remove trailing whitespace

Shut off automatic removal in my editor to avoid touching other
people's code, but that resulted in not auto-fixing it in my own
code.

Randalphwa commented:
I rebased from upstream before amending the PR, but while I was running doxygen to verify the changes to xrc_format.h, another commit got added to master. That's showing up as a merge commit in my branch, and that's beyond by git knowledge to prevent. It's correct with the commit in the master branch, but messes with the history. I'm not sure what else to do on my end (other than being certain in the future to rebase just prior to pushing, rather than before testing).

vadz reviewed (commented):
Sorry, I'm not sure what is the problem with Git, but, in any case, I can merge/rebase myself as/if needed. I'd just like to ask you to make the change below (and test that it works as expected by trying to use a wrong parameter value) and then we can merge it.

TIA!

Randalphwa committed:
Report if invalid checked param is used

Since checked is now an integer, need to let user know that
-1 is not a valid option.

Randalphwa commented:
Good point. I made the changes and verified that -1 generates the expected warning in a debug build.

vadz commented:
Great, thanks, will merge soon.
